File: solution300.java


public class solution300 {
    public int lengthOfLIS(int[] nums) {
        if (nums == null || nums.length < 1) {
            return 0;
        }
        int[] dp = new int[nums.length];
        int res = 1;
        dp[0] = 1;
        for (int i = 1; i < dp.length; i++) {
            int maxLength = 0;
            for (int j = i - 1; j >= maxLength; j--) {
                if (nums[j] <= nums[i]) {
                    maxLength = dp[j] > maxLength ? dp[j] : maxLength;
                }
            }
            dp[i] = maxLength + 1;
            res = dp[i] > res ? dp[i] : res;
        }
        return res;
    }
}

File: solution300.go
package leetcode

func lengthOfLIS(nums []int) int {
	if len(nums) < 1 {
		return 0
	}
	var dp []int = make([]int, len(nums))
	dp[0] = 1
	var res int = 1
	for i := 1; i < len(nums); i++ {
		maxLen := 0
		for j := i - 1; j >= maxLen; j-- {
			if nums[j] < nums[i] && dp[j] > maxLen {
				maxLen = dp[j]
			}
		}
		dp[i] = maxLen + 1
		if dp[i] > res {
			res = dp[i]
		}
	}
	return res
}
